### PR TITLE
Fix inaccuracies in vis_heap_chunks()

### LINKED ISSUE #706: vis_heap_chunks() output is inaccurate
The following screenshot shows the vis_heap_chunks() command output when introspecting the default heap, which has 3 0x20-sized chunks allocated on it.

vis_heap_chunks errors

The first command incorrectly shows the quadword at address 0x405040 as belonging to the 3rd (magenta) chunk, when it actually belongs to the 2nd (cyan) chunk. This issue is only present in the dev branch. The output also shows the quadword at address 0x405060 as belonging to the top chunk, when it belongs to the 3rd chunk. This issue is present in both dev & stable branches.

The second command does not show the last quadword of the yellow chunk's user data, despite it being correctly marked as belonging to the yellow chunk in the first command output. This issue is present in both dev & stable branches.

### PR DESCRIPTION
**Pull request by CptGibbon**
The `vis_heap_chunks` command incorrectly determines the boundaries between chunks near to a top chunk. For example, in the image below there are 3 identical, allocated chunks of minimum size which `vis_heap_chunks` wrongly shows as having different shapes. The 2nd invocation has not printed the trailing quadword of the 1st chunk, despite it being correctly marked as belonging to the chunk in the 1st output.

image

This PR corrects the boundary calculation of chunks adjacent to a top chunk, it also uses a list of boundary addresses to do so, rather than a dictionary of every quadword in each printed chunk. The image below shows `vis_heap_chunks` output after the patch has been applied, each chunk looks identical and the 2nd invocation prints the last quadword of the 1st chunk.

image

The colours change between screenshots because the original implementation subtracted 1 from a `color_func_idx` variable that was already 0 before building the output buffer. Fixes #706

### CODE REVIEW

**Comment by disconnect3d** [comment]
@GrosQuildu or @stnevans any chance one of u could review this?

File: pwndbg/commands/heap.py [hunk 1]
```diff
@@ -495,51 +495,51 @@ def find_fake_fast(addr, size):
...
+
+    for _ in range(count + 1):
+        # Don't read beyond the heap mapping if --naive or corrupted heap.
+        if cursor not in heap_region:
+            chunk_delims.append(heap_region.end)
+            break
+
+        size_field = pwndbg.memory.u(cursor + ptr_size)
+        real_size = size_field & ~main_heap.malloc_align_mask
+        prev_inuse = size_field & pwndbg.constants.ptmalloc.PREV_INUSE
```
Comment (GrosQuildu) [inline pwndbg/commands/heap.py]:
```suggestion
        prev_inuse = main_heap.chunk_flags(size_field)[0]
```

File: pwndbg/commands/heap.py [hunk 2]
```diff
@@ -495,51 +495,51 @@ def find_fake_fast(addr, size):
...
-
+    heap_region = main_heap.get_heap_boundaries(address)
     top_chunk = get_top_chunk_addr()
- 
-
-    unpack = pwndbg.arch.unpack
 
-    cells_map = {}
-    chunk_id = 0
     ptr_size = pwndbg.arch.ptrsize
```
Comment (GrosQuildu) [inline pwndbg/commands/heap.py]: Just for quality, not really important.
```suggestion
    ptr_size = main_heap.size_sz
```

File: pwndbg/commands/heap.py [hunk 3]
```diff
@@ -495,51 +495,51 @@ def find_fake_fast(addr, size):
...
+        # Don't repeatedly operate on the same address (e.g. chunk size of 0).
+        if cursor in chunk_delims or cursor + ptr_size in chunk_delims:
+            break
+
+        if prev_inuse:
+            chunk_delims.append(cursor + ptr_size)
+        else:
+            chunk_delims.append(cursor)
+
+        if (cursor == top_chunk and not naive) or (cursor == heap_region.end - 0x10):
```
Comment (GrosQuildu) [inline pwndbg/commands/heap.py]: 0x10 - magic constant, maybe better will be:
```suggestion
        if (cursor == top_chunk and not naive) or (cursor == heap_region.end - ptr_size*2):
```

File: pwndbg/commands/heap.py [hunk 4]
```diff
@@ -563,30 +562,33 @@ def vis_heap_chunks(address=None, count=None, naive=None):
     out = ''
     asc = ''
     labels = []
+    cursor = int(address) if address else heap_region.start
+
+    for c, stop in enumerate(chunk_delims):
+        color_func_idx = c % len(color_funcs)
+        color_func = color_funcs[color_func_idx]
```
Comment (GrosQuildu) [inline pwndbg/commands/heap.py]:
```suggestion
        color_func = color_funcs[c % len(color_funcs)]
```

File: pwndbg/commands/heap.py [hunk 5]
```diff
@@ -563,30 +562,33 @@ def vis_heap_chunks(address=None, count=None, naive=None):
...
-        if printed % 2 == 0:
-            out += "\n0x%x" % addr
+        while cursor != stop:
+            if printed % 2 == 0:
+                out += "\n0x%x" % cursor
 
-        cell = unpack(pwndbg.memory.read(addr, ptr_size))
-        cell_hex = '\t0x{:0{n}x}'.format(cell, n=ptr_size*2)
+            cell = pwndbg.memory.u(cursor)
+            cell_hex = '\t0x{:0{n}x}'.format(cell, n=ptr_size*2)
```
Comment (GrosQuildu) [inline pwndbg/commands/heap.py:549]: Does pwndbg have some uniform way to print stuff like that? That is: address, memory, content, and labels, kinda like hexdump. @disconnect3d 

Probably not?

**Comment by GrosQuildu** [comment]
Can we make it works with non-heap regions? Running `vis_heap_chunks` on arbitrary address may be useful when you are constructing fake chunks on i.e. stack. Currently its not possible because [`main_heap.get_heap_boundaries(address)`](https://github.com/pwndbg/pwndbg/blob/7f963fed0e8f54de93101a33ce30e66d497fb058/pwndbg/commands/heap.py#L508) crashes.

**Review by GrosQuildu — CHANGES_REQUESTED** [review]

File: pwndbg/commands/heap.py (hunk 5 above)
Comment (disconnect3d) [reply pwndbg/commands/heap.py:549]: I think not

**Comment by disconnect3d** [comment]
@CptGibbon ping :)

**Comment by CptGibbon** [comment]
> @CptGibbon ping :)

I'm on it 👍

**Comment by CptGibbon** [comment]
> Can we make it works with non-heap regions? Running `vis_heap_chunks` on arbitrary address may be useful when you are constructing fake chunks on i.e. stack. Currently its not possible because [`main_heap.get_heap_boundaries(address)`](https://github.com/pwndbg/pwndbg/blob/7f963fed0e8f54de93101a33ce30e66d497fb058/pwndbg/commands/heap.py#L508) crashes.

Sure, I've addressed this in 62f1180.

**Comment by GrosQuildu** [comment]
Seems ok, thanks.

@disconnect3d can merge that one.

**Review by GrosQuildu — APPROVED** [review]
OK

**Pull request merged**